CMSIS/NN/Source/PoolingFunctions/arm_avgpool_s8.c

```diff
@@ -228,51 +248,101 @@ arm_status arm_avgpool_s8(const int dim_src_height,
                           int8_t *dst)
 {
 
-/* Reference C code adapted from CMSIS-NN arm_avepool_q7_HWC.
- */
-    (void)bufferA;
-    int16_t   i_ch_in, i_x, i_y;
-    int16_t   k_x, k_y;
+#if defined (ARM_MATH_DSP)
+
+  /* Run the following code for Cortex-M4 and Cortex-M7
+   */
+  int32_t k_x, k_y, i_x, i_y;
 
-    for (i_y = 0; i_y < dim_dst_height; i_y++)
+  for (i_y = 0; i_y < dim_dst_height; i_y++)
+  {
+    for (i_x = 0; i_x < dim_dst_width; i_x++)
     {
-        for (i_x = 0; i_x < dim_dst_width; i_x++)
+      /* Condition for kernel start dimension: (base_idx_<x,y> + kernel_<x,y>_start) >= 0 */
+      const int32_t base_idx_y = (i_y * stride_height) - padding_height;
+      const int32_t base_idx_x = (i_x * stride_width) - padding_width;
+      const int32_t kernel_y_start = MAX(0, -base_idx_y);
+      const int32_t kernel_x_start = MAX(0, -base_idx_x);
+
+      /* Condition for kernel end dimension: (base_idx_<x,y> + kernel_<x,y>_end) < dim_src_<width,height> */
+      const int32_t kernel_y_end = MIN(dim_kernel_height, dim_src_height - base_idx_y);
+      const int32_t kernel_x_end = MIN(dim_kernel_width, dim_src_width - base_idx_x);
+
+      int count = 0;
+
+      for (k_y = kernel_y_start; k_y < kernel_y_end; k_y++)
+      {
+        for (k_x = kernel_x_start; k_x < kernel_x_end; k_x++)
         {
-            for (i_ch_in = 0; i_ch_in < ch_src; i_ch_in++)
+          q7_t *start = src + ch_src * (k_x + base_idx_x + (k_y + base_idx_y) * dim_src_width);
+
+          if (count == 0)
+          {
+            arm_q7_to_q15_no_shift(start, bufferA, ch_src);
+          }
+          else
+          {
+            arm_nn_accumulate_q7_to_q15(bufferA, start, ch_src);
```

Comment:
I think we should document and add-restriction  the possible overflow case here as we are accumulating on to 16 bits only. Or accumulate on 32 bits.. We accumulate on to 32 bits in MVE and non-DSP case..

Reply:
Ok, let's accumulate on to 32 bits then